myset.h

```diff
@@ -0,0 +1,156 @@
+            right = nullptr;
+            minim = this;
+            par = nullptr;
+            prior = uniform(rng);
+        }
+    };
+    void renew(Node* &n){
+        if (!n)
+            return;
+
+        n->minim = n->left != nullptr ? n->left->minim : n;
+        if (n -> left)
+            n -> left -> par = n;
+        if (n -> right)
+            n -> right -> par = n;
+    }
+    std::tuple <Node*, Node*> split(Node *tree,const T &key){
+        Node *left = nullptr, *right = nullptr;
+        if (!tree)
+            return std::make_tuple(left, right);
+        auto cmp = Compare();
+        if (cmp(tree -> key, key)) { // tree->key < key
+            left = tree;
+            std::tie(left -> right, right) = split(tree->right, key);
+        } else{ // tree->key > key
+            right = tree;
+            std::tie(left, right->left) = split(tree -> left, key);
+        }
+        renew(left);
+        renew(right);
+        if (left)
+            left -> par = nullptr;
+        if (right)
+            right -> par = nullptr;
+        return std::make_tuple(left, right);
+    }
+    Node* merge(Node *t1, Node *t2){
+        if (!t1)
+            return t2;
+        if (!t2)
+            return t1;
+
+        if (t1 -> prior > t2 -> prior){
+            t1 -> right = merge(t1 -> right, t2);
+            renew(t1);
+            return t1;
+        }
+        t2 -> left = merge(t1, t2 -> left);
+        renew(t2);
+        return t2;
+    }
+    void push_min(Node *n){
+        if  (!n)
+            return;
+        n-> minim = n -> left -> minim;
+        push_min(n -> par);
+    };
+    void replace_with_right(Node *t){ // consider right exists
+
+        if (t -> par){
+            t -> par -> left = t -> right;
+            t -> right -> par = t -> par;
+        } else{
+            root = t -> right;
+            root -> par = nullptr;
+        }
+    }
+    void dealloc_entity(Node *n){
+        table.erase(n->key);
+        delete n;
+    }
+    void upd_min(Node *mi){
+        Node* t = mi;
+        for(;;) {
+            if (t->count > 0) {
+                t->minim = t;
+                push_min(t->par);
+                return;
+            }
+            if (t->right) {
+                replace_with_right(t);
+                Node *m = t->right->minim;
+                dealloc_entity(t);
+                t = m;
+            } else {
+                if (t->par) {
+                    t->par->left = nullptr;
+                    Node *par = t->par;
+                    dealloc_entity(t);
+                    t = par;
+                } else {
+                    dealloc_entity(t);
+                    root = nullptr;
+                    root -> minim = root;
+                    return;
+                }
+            }
+        }
+    }
+    Node *root = nullptr;
+    std::unordered_map <T, Node*> table;
+public:
+    void erase(const T x){
+        if (table.contains(x)){
+            table[x]->count--;
```

Comment:
You search for aan element in the table twice, it is ok with unordered_map, but if it is map, it will be not so good. Think about how you can make it using `find`.